examples/jsm/loaders/KTX2Loader.js

```diff
@@ -272,7 +272,24 @@ class KTX2Loader extends Loader {
 
 		if ( container.vkFormat !== VK_FORMAT_UNDEFINED ) {
 
-			return createDataTexture( container );
+			const mipmaps = [];
+			const pendings = [];
+
+			for ( let levelIndex = 0; levelIndex < container.levels.length; levelIndex ++ ) {
+
+				pendings.push( createDataTexture( container, levelIndex ).then( function ( dataTexture ) {
+
+					mipmaps[ levelIndex ] = dataTexture;
+
+				} ) );
+
+			}
+
+			await Promise.all( pendings );
+
+			const texture = mipmaps.shift();
+			texture.mipmaps = mipmaps;
```

Comment:
It doesn't seem right that DataTexture#mipmaps field contains an array of other DataTextures... but I'm not sure what is correct here. Do we already support mipmaps on DataTexture, and does this PR render as expected?

Reply:
>Do we already support mipmaps on DataTexture

I wasn't able to find other code in 3js or any 3js examples that reference `DataTexture.mipmaps`. The [DataTexture docpage](https://threejs.org/docs/#api/en/textures/DataTexture) doesn't mention the `mipmap` field. Of course, the [Texture docpage](https://threejs.org/docs/#api/en/textures/Texture) does, but it's marked as "optional". What specific requirements were you thinking of re: "support"?

>does this PR render as expected?

Full disclosure: For my current use case, I actually wanted to load KTX2s into CubeTextures. After manually post-processing (splitting the data into six faces, etc.) all the DataTextures into CubeTextures (including the ones in `.mipmaps`), they do render correctly. I was hoping to find a suitable place to upstream that into 3js afterwards too.

Reply:
> What specific requirements were you thinking of re: "support"?

KTX2Loader should return a texture that can be used by the renderer without further processing. Mipmaps are expected to have sequential sizes in powers of two, so storing faces of a cubemap in mips doesn't seem quite right.

KTX2 has non-mipmap features that would be relevant here: [array textures and cube textures](https://github.com/donmccurdy/KTX-Parse/blob/c8dd388a25a2ba80279d454b7b7e11f436d6ae0a/src/container.ts#L40-L48). Depending on how your KTX2 file is formatted, we could interpet it as a [THREE.DataArrayTexture](https://threejs.org/docs/?q=texture#api/en/textures/DataArrayTexture), or perhaps a CubeTexture if the KTX2 file's `faceCount` is set. 

@Mugen87 is there (or should there be) a Cube- equivalent of DataTexture? 

@emmanueljl could you share 1-2 example KTX2 files that you intend to support? Having samples will be important for us to maintain this codepath. Thanks!

Reply:
> @Mugen87 is there (or should there be) a Cube- equivalent of DataTexture?

It's best to use the approach of `HDRCubeTextureLoader` which assigns instances of `DataTexture` to its `images` array property.

Reply:
Ok! In that case I believe KTX2Loader should return a CubeTexture if `faceCount` is >1, or `DataArrayTexture` for array textures. Feel free to implement just one or the other for this PR of course.

Reply:
Thanks for the pointers! I brought up the assembly of `CubeTexture`s from the loaded `DataTexture`s mainly to say that this loading code works as expected, and renders properly in my context. Let's focus on this loader PR first and pick back up the `CubeTexture` discussion in a future PR.

That said, for the last few days I've been trying to put together [this 3js sample](https://emmanuellee.com/three.js/examples/webgl_loader_texture_ktx2.html) along with [this synthetic KTX2](https://emmanuellee.com/three.js/examples/textures/compressed/mipmaps_sample.ktx2) (contains mips of different solid colors) to test rendering of the `DataTexture` that's created from a KTX2 with mipmaps, but I've run into a few issues/questions with texture uploading.

1. When uploading a `DataTexture` with mipmaps, [this line](https://github.com/mrdoob/three.js/blob/805f1865a7d3a51086c3e5f1037680e6e8bae1cf/src/renderers/webgl/WebGLTextures.js#L825) incorrectly looks for `width`, `height`, `data` in `mipmap` rather than in `mipmap.source` or `mipmap.image`. Is this a bug? I wonder if this codepath simply isn't used much. Below, when uploading a `DataTexture` without mipmaps, [this line](https://github.com/mrdoob/three.js/blob/805f1865a7d3a51086c3e5f1037680e6e8bae1cf/src/renderers/webgl/WebGLTextures.js#L847) correctly references `image.width`, `image.height`, `image.data`. However, even after fixing that line to use `mipmap.image`, I get: `GL_INVALID_VALUE: Texture dimensions must all be greater than zero.`

2. It seems 3js expects a texture's base mip to be included as the first entry in its own `mipmaps` array. Is that right?

Reply:
>  looks for width, height, data in mipmap rather than in mipmap.source or mipmap.image ...

A texture's mipmaps are not textures, they are just plain objects containing `width`, `height`, and `data` entries for each mip. This code could be helpful as a reference, although it deals with compressed textures:

https://github.com/mrdoob/three.js/blob/49f7bf498e1b37a65b215a17820fb536f44ab2a1/examples/jsm/loaders/KTX2Loader.js#L518

Also – it's not possible for two mips of a [2D, non-array] texture to have the same dimensions. I see this sample has 6 mips ("levels") with sizes [32px, 16px, 8px, 4px, 2px, 1px], from ktxinfo output:

```
pixelWidth: 32
pixelHeight: 32
pixelDepth: 0
layerCount: 0
faceCount: 1
levelCount: 6
```

Is that what you're expecting to find? The KTX2 file would need to be authored slightly differently to contain 6 faces of the same resolution for a cubemap.

Reply:
>A texture's mipmaps are not textures, they are plain objects containing width, height, and data entries for each mip.

I see! Then the way my PR loads levels is wrong. I'll fix that up.

>it's not possible for two mips of a [2D, non-array] texture to have the same dimensions.

By this, do you mean that the correct resulting `DataTexture` from loading my KTX2 should be 32px and then have `mipmaps` of [16px, 8px, 4px, 2px, 1px] (as opposed to `mipmaps` of [**32px,** 16px, 8px, 4px, 2px, 1px])?

>Is that what you're expecting to find? The KTX2 file would need to be authored slightly differently to contain 6 faces of the same resolution for a cubemap.

Yes, they are just POT squares of different solid colors to test with. Let's worry about cubemapping in a future PR.

Reply:
> By this, do you mean ... mipmaps of [16px, 8px, 4px, 2px, 1px] (as opposed to mipmaps of [32px, 16px, 8px, 4px, 2px, 1px])

Oops, I just meant that we can't store a cube texture as mips like [32px, 32px, 32px, ...], the sizes have to descend. Sounds like we're on the same page there. The mipmaps should contain all of the texture's data, the 0th mip is the full-resolution data.

> Let's worry about cubemapping in a future PR.

Ok, sounds good!